Mathlib/GroupTheory/SpecificGroups/KleinFour.lean

```diff
@@ -39,74 +40,195 @@ produces the third one.
+/-! # Properties of groups with exponent two -/
 
-namespace KleinFour
-
-open DihedralGroup Equiv
-
-/-- Element `a` of Klein four-group. -/
-def a : KleinFour := Multiplicative.ofAdd (0, 1)
-
-/-- Element `b` of Klein four-group. -/
-def b : KleinFour := Multiplicative.ofAdd (1, 0)
-
-/-- Element `c` of Klein four-group. -/
-def c : KleinFour := Multiplicative.ofAdd (1, 1)
-
-/-- Klein four-group is a group of order 4. -/
-theorem card : Fintype.card KleinFour = 4 :=
-  rfl
-
-/-- Klein four-group is a group of order 4. -/
-theorem nat_card : Nat.card KleinFour = 4 := by
-  simp (config := {decide := true}) only [Nat.card_eq_fintype_card]
-
-@[simp] theorem a_sq : a ^ 2 = 1 :=
-  rfl
-
-@[simp] theorem b_sq : b ^ 2 = 1 :=
-  rfl
-
-@[simp] theorem c_sq : c ^ 2 = 1 :=
-  rfl
-
-@[simp] theorem orderOf_a : orderOf a = 2 :=
-  orderOf_eq_prime a_sq (by decide)
-
-@[simp] theorem orderOf_b : orderOf b = 2 :=
-  orderOf_eq_prime b_sq (by decide)
-
-@[simp] theorem orderOf_c : orderOf c = 2 :=
-  orderOf_eq_prime c_sq (by decide)
-
-@[simp high]
-theorem exponent : Monoid.exponent KleinFour = 2 := by
-  simp [AddMonoid.exponent_prod]
-
-/-- Klein four-group is a non-cyclic group. -/
-theorem notIsCyclic : ¬ IsCyclic KleinFour := by
-  intro h
-  have h₁ := IsCyclic.iff_exponent_eq_card.mp h
-  rw [exponent, card] at h₁
-  contradiction
-
-/-- Klein four-group is isomorphic to the Dihedral group of order 4. -/
-@[simps]
-def mulEquivDihedralGroupTwo : KleinFour ≃* DihedralGroup 2 where
-  toFun := fun
-    | (0, 0) => 1
-    | (0, 1) => sr 1
-    | (1, 0) => r 1
-    | (1, 1) => sr 1 * r 1
-  invFun := fun
-    | 1 => (0, 0)
-    | sr 1 => (0, 1)
-    | r 1 => (1, 0)
-    | sr 1 * r 1 => (1, 1)
-  left_inv := by decide
-  right_inv := by decide
-  map_mul' := by simp (config := {decide := true})
-
-end KleinFour
+section ExponentTwo
+
+variable {G : Type*} [Group G]
+
+/-- In a group of exponent two, every element is its own inverse. -/
+@[to_additive]
+lemma inv_eq_self_of_exponent_two (hG : Monoid.exponent G = 2) (x : G) :
+    x⁻¹ = x :=
+  inv_eq_of_mul_eq_one_left <| pow_two (a := x) ▸ hG ▸ Monoid.pow_exponent_eq_one x
+
+/-- If an element in a group has order two, then it is its own inverse. -/
+@[to_additive]
+lemma inv_eq_self_of_orderOf_eq_two {x : G} (hx : orderOf x = 2) :
+    x⁻¹ = x :=
+  inv_eq_of_mul_eq_one_left <| pow_two (a := x) ▸ hx ▸ pow_orderOf_eq_one x
+
+@[to_additive]
+lemma orderOf_eq_two_iff (hG : Monoid.exponent G = 2) {x : G} :
+    orderOf x = 2 ↔ x ≠ 1 :=
+  ⟨by rintro hx rfl; norm_num at hx, orderOf_eq_prime (hG ▸ Monoid.pow_exponent_eq_one x)⟩
+
+/-- In a group of exponent two, all elements commute. -/
+@[to_additive]
+lemma mul_comm_of_exponent_two (hG : Monoid.exponent G = 2) (x y : G) :
+    x * y = y * x := by
+  simpa only [inv_eq_self_of_exponent_two hG] using mul_inv_rev x y
+
+/-- Any group of exponent two is abelian. -/
+@[to_additive (attr := reducible) "Any additive group of exponent two is abelian."]
```

Comment:
I think that all the result up to here can go to a more basic file (where the exponent is defined?).

Reply:
I considered that, but I felt it would be better here so as not to clutter that file with facts about exponent two. Most likely, if you want results about exponent two, you'll be interested in the Klein four-group anyway.

If you still think I should move it that's fine.